ProjectLighthouse/Controllers/ReviewController.cs

```diff
@@ -74,5 +81,177 @@ public async Task<IActionResult> DPadRate(int slotId, [FromQuery] int rating)
+                
+            string inner = Enumerable.Aggregate(reviews, string.Empty, (current, review) => {
+                RatedLevel? ratedLevel = this.database.RatedLevels.FirstOrDefault(r => r.SlotId == review.SlotId && r.UserId == user.UserId);
+                RatedReview? ratedReview = this.database.RatedReviews.FirstOrDefault(r => r.ReviewId == review.ReviewId && r.UserId == user.UserId);
+                return current + review.Serialize(ratedLevel, ratedReview);
+            });
+
+            string response = LbpSerializer.TaggedStringElement("reviews", inner, new Dictionary<string, object>
+                    {
+                        {
+                            "hint_start", pageStart
+                        },
+                        {
+                            "hint", reviews.Last().Timestamp // Seems to be the timestamp of oldest
+                        },
+                    });
+
+            return this.Ok(response);
+        }
+
+        [HttpPost("rateReview/user/{slotId:int}/{username}")]
+        public async Task<IActionResult> RateReview(int slotId, string username, [FromQuery] int rating = 0) {
+            User? user = await this.database.UserFromRequest(this.Request);
+            if (user == null) return this.StatusCode(403, "");
+
+            User? reviewer = await this.database.Users.FirstOrDefaultAsync(u => u.Username == username);
+            if (reviewer == null) return this.StatusCode(403, "");
+
+            Review? review = await this.database.Reviews.FirstOrDefaultAsync(r => r.SlotId == slotId && r.ReviewerId == reviewer.UserId);
+            if (review == null) return this.StatusCode(403, "");
+
+            RatedReview? ratedReview = await this.database.RatedReviews.FirstOrDefaultAsync(r => r.ReviewId == review.ReviewId && r.UserId == user.UserId);
+            if (ratedReview == null)
+            {
+                ratedReview = new RatedReview();
+                ratedReview.ReviewId = review.ReviewId;
+                ratedReview.UserId = user.UserId;
+                ratedReview.Thumb = 0;
+                this.database.RatedReviews.Add(ratedReview);
+            }
+
+            ratedReview.Thumb = Math.Max(Math.Min(1, rating), -1);
+
+            await this.database.SaveChangesAsync();
+
+            return this.Ok();
+        }
+
+        [HttpPost("deleteReview/user/{slotId:int}/{username}")]
+        public async Task<IActionResult> DeleteReview(int slotId, string username) {
+            User? reviewer = await this.database.Users.FirstOrDefaultAsync(u => u.Username == username);
+            if (reviewer == null) return this.StatusCode(403, "");
+
+            Review? review = await this.database.Reviews.FirstOrDefaultAsync(r => r.SlotId == slotId && r.ReviewerId == reviewer.UserId);
+            if (review == null) return this.StatusCode(403, "");
+
+            review.Deleted = true;
+            review.DeletedBy = "level_author"; // other value is "moderator"
```

Comment:
Can we have an enum for this for simplicity's sake?

Reply:
Well, I also learned today that C# can't have string enums so this is close enough, works practically the same way. https://github.com/LBPUnion/project-lighthouse/pull/62/commits/ece00c218ede04b4226bf45bcc27fc4449c4e557

I opted not to namespace it with reviews specifically because comments (both profile and level) can be deleted and I thought it would be a bit redundant to have separate types for each.